gpu/containers/src/device_memory.cpp

```diff
@@ -174,11 +174,43 @@ void pcl::gpu::DeviceMemory::upload(const void *host_ptr_arg, std::size_t sizeBy
     cudaSafeCall( cudaDeviceSynchronize() );
 }
 
+bool
+pcl::gpu::DeviceMemory::upload(void* host_ptr_arg,
+                               std::size_t device_begin_byte_offset,
+                               std::size_t num_bytes)
+{
+  const void* const begin = static_cast<char*>(data_) + device_begin_byte_offset;
+  const char* const upload_end = static_cast<const char*>(begin) + num_bytes;
+  const char* const array_end = static_cast<char*>(data_) + sizeBytes_;
+  if (upload_end > array_end) {
+      return false;
+  }
+  cudaSafeCall(cudaMemcpy(host_ptr_arg, begin, num_bytes, cudaMemcpyHostToDevice));
+  cudaSafeCall(cudaDeviceSynchronize());
+  return true;
+}
+
 void pcl::gpu::DeviceMemory::download(void *host_ptr_arg) const
 {    
     cudaSafeCall( cudaMemcpy(host_ptr_arg, data_, sizeBytes_, cudaMemcpyDeviceToHost) );
     cudaSafeCall( cudaDeviceSynchronize() );
-}          
+}
+
+bool
+pcl::gpu::DeviceMemory::download(void* host_ptr_arg,
+                                 std::size_t device_begin_byte_offset,
+                                 std::size_t num_bytes) const
+{
+  const void* const begin = static_cast<char*>(data_) + device_begin_byte_offset;
+  const char* const download_end = static_cast<const char*>(begin) + num_bytes;
+  const char* const array_end = static_cast<char*>(data_) + sizeBytes_;
+  if (download_end > array_end) {
```

Comment:
Correct me if I'm wrong, but I think `(device_begin_byte_offset + num_bytes) > sizeBytes_` should have the same effect, and is IMO a bit more readable (same for upload)

Reply:
This is great! I think you are correct, and it's much more readable. I will think about its correctness a bit more and apply the changes. Thanks for the suggestion!

Reply:
This was an excellent comment - thank you, Markus! Your remark also made me realize that the "const-ness" of the function arguments was wrong. I hope/think it is correct now. Again, thank you!